File: backends/typescript/bridge.ts
// py<->node interop bridge, consumer (Node / cordis) side.
//
// Mirror of backends/python/bridge.py's proxy half: a cordis component that
// provides a key by forwarding calls to a Unix-socket stub running in another
// process (here, the Python provider from demo/bridge_pypy.py). To the
// consumer, `ctx.db` is an ordinary provider; the language boundary is invisible.
//
// The crossed method here, Database.execute, is SYNCHRONOUS, which `revl audit`
// classifies as address-space-bound: a sync call over a socket is chatty. We
// honor that literally. Each call is a blocking round-trip (execFileSync a
// one-shot client), so the demo pays exactly the cost the audit predicts,
// while an `async fn` service would proxy without blocking. A persistent
// monitor connection turns provider death into provision withdrawal (R2/R3).

import type { Context } from 'cordis'
import { execFileSync } from 'node:child_process'
import fs from 'node:fs'
import net from 'node:net'
import tls from 'node:tls'

// --- seam endpoints: a local UDS (default) or a network TCP + mTLS seam -------
//
// The node client mirrors backends/python/bridge.py's `Endpoint`/`TlsConfig`
// (docs/network-placement.md): a bare socket-path string is a local UDS (full
// back-compat), and a `{ host, port, tls }` object is a network seam reached
// over TCP wrapped in mutual TLS. The provider (py, item 56) demands the
// consumer's certificate (`CERT_REQUIRED`) and the consumer verifies the
// provider's against the same CA — a network seam is "the two processes holding
// CA-signed certs", not "whoever can reach the port". The canonical value
// codec, the seam deadline (item 54) and reactive withdrawal on peer death all
// apply unchanged over TCP; only the address family and the TLS wrap differ.

export interface TlsPaths {
  cert: string
  key: string
  ca: string
  identity: string
  server_hostname?: string
}

export interface NetworkEndpoint {
  host: string
  port: number
  tls: TlsPaths
}

/** A seam target: a UDS path (legacy/local) or a network TCP+mTLS endpoint. */
export type SeamTarget = string | NetworkEndpoint

function isNetwork(target: SeamTarget): target is NetworkEndpoint {
  return typeof target !== 'string' && target != null && typeof target.host === 'string'
}

/** A seam call that outlived its deadline: the provider is neither answering nor
 *  gone — it hung. The node mirror of backends/python/bridge.py's `SeamDeadline`
 *  (a distinguishable fault, disjoint from a peer death and from a remote
 *  error), so a caller — or a test — can tell a hang apart by kind. */
export class SeamDeadlineError extends Error {
  readonly key: string
  readonly method: string
  readonly deadlineMs: number
  constructor(key: string, method: string, deadlineMs: number) {
    super(`revl: seam call ${key}.${method} exceeded its ${deadlineMs}ms deadline`)
    this.name = 'SeamDeadlineError'
    this.key = key
    this.method = method
    this.deadlineMs = deadlineMs
  }
}

// One-shot client: connect (UDS or TCP+mTLS), send one request line, print the
// reply line, exit. Run as `node -e`, fed the target/request/deadline via the
// environment. A wedged provider trips the in-child deadline timer and the child
// prints a `{ seamDeadline: true }` reply rather than hanging; a TCP refusal
// (provider still coming up) is retried, so start order stays irrelevant.
const CLIENT_SRC = `
const net = require('node:net')
const tls = require('node:tls')
const fs = require('node:fs')
const mode = process.env.BRIDGE_MODE
const deadlineMs = process.env.BRIDGE_DEADLINE_MS ? Number(process.env.BRIDGE_DEADLINE_MS) : 0
let attempts = 0
let timer = null
let secured = false
function done(obj) {
  if (timer) clearTimeout(timer)
  process.stdout.write(JSON.stringify(obj))
  process.exit(0)
}
if (deadlineMs > 0) {
  timer = setTimeout(() => done({ ok: false, seamDeadline: true, error: 'seam call exceeded ' + deadlineMs + 'ms deadline' }), deadlineMs)
}
function dial() {
  let s
  if (mode === 'tcp') {
    s = tls.connect({
      host: process.env.BRIDGE_HOST,
      port: Number(process.env.BRIDGE_PORT),
      ca: fs.readFileSync(process.env.BRIDGE_CA),
      cert: fs.readFileSync(process.env.BRIDGE_CERT),
      key: fs.readFileSync(process.env.BRIDGE_KEY),
      servername: process.env.BRIDGE_SERVERNAME || process.env.BRIDGE_HOST,
    })
    s.on('secureConnect', () => { secured = true; s.write(process.env.BRIDGE_REQ + '\\n') })
  } else {
    s = net.connect(process.env.BRIDGE_SOCK)
    s.on('connect', () => s.write(process.env.BRIDGE_REQ + '\\n'))
  }
  let buf = ''
  s.on('data', (d) => {
    buf += d
    const i = buf.indexOf('\\n')
    if (i >= 0) { s.destroy(); done(JSON.parse(buf.slice(0, i))) }
  })
  s.on('error', () => {
    s.destroy()
    // A TLS handshake failure (bad cert / wrong CA) is terminal — never retry it
    // away as if the provider were merely still coming up.
    if (mode === 'tcp' && secured) { done({ ok: false, error: 'bridge tls stream error' }) }
    if (++attempts > 100) { done({ ok: false, error: 'bridge connect failed' }) }
    else setTimeout(dial, 50) // provider may still be coming up
  })
}
dial()
`

// Canonical ADT/Result wire codec (docs/interop-bridge.md "Canonical value
// encoding"). The TS runtime ADT form is `{ kind, value? }` (records are plain
// objects, Opt is bare value | undefined). On the wire a tagged value is
// `{ "$kind": Case, "$value"?: payload }`; the `$kind` marker is what separates
// it from a record. So encode renames kind/value -> $kind/$value, decode the
// reverse, both recursively; records/arrays/scalars/null pass through.

function isNativeAdt(o: Record<string, unknown>): boolean {
  return typeof o.kind === 'string' && Object.keys(o).every((k) => k === 'kind' || k === 'value')
}

/** `Int` is a `bigint` on this tier, and the wire says `Int` crosses as a JSON
 *  *number* (docs/interop-bridge.md, "Canonical value encoding"). JSON.stringify
 *  THROWS on a bigint, so it has to convert here — and it refuses rather than
 *  rounding when the value is outside the range a JSON number holds exactly.
 *  Rounding at a seam is the silent precision loss this tier's BigInt port
 *  exists to remove; a 64-bit `Int` that does not fit the wire is a limit of
 *  the wire, and it should say so rather than hand over a different number. */
function bigintToWire(v: bigint): number {
  if (v > BigInt(Number.MAX_SAFE_INTEGER) || v < BigInt(Number.MIN_SAFE_INTEGER)) {
    throw new RangeError(
      `revl: Int ${v} is outside the range a JSON number represents exactly ` +
        `(docs/interop-bridge.md encodes Int as a number)`,
    )
  }
  return Number(v)
}

/** Bigints in an outbound argument, and nothing else. The request path does
 *  not run the ADT codec below — changing that is a cross-language decision,
 *  not this tier's — so this only replaces a guaranteed throw. */
function encodeBigInts(v: unknown): unknown {
  if (typeof v === 'bigint') return bigintToWire(v)
  if (Array.isArray(v)) return v.map(encodeBigInts)
  if (v && typeof v === 'object' && !(v instanceof Map)) {
    const o = v as Record<string, unknown>
    const rec: Record<string, unknown> = {}
    for (const k of Object.keys(o)) rec[k] = encodeBigInts(o[k])
    return rec
  }
  return v
}

export function encodeValue(v: unknown): unknown {
  if (typeof v === 'bigint') return bigintToWire(v)
  if (Array.isArray(v)) return v.map(encodeValue)
  if (v && typeof v === 'object' && !(v instanceof Map)) {
    const o = v as Record<string, unknown>
    if (isNativeAdt(o)) {
      const out: Record<string, unknown> = { $kind: o.kind }
      if ('value' in o) out.$value = encodeValue(o.value)
      return out
    }
    const rec: Record<string, unknown> = {}
    for (const k of Object.keys(o)) rec[k] = encodeValue(o[k])
    return rec
  }
  return v
}

export function decodeValue(v: unknown): unknown {
  if (Array.isArray(v)) return v.map(decodeValue)
  if (v && typeof v === 'object') {
    const o = v as Record<string, unknown>
    if (typeof o.$kind === 'string') {
      const out: Record<string, unknown> = { kind: o.$kind }
      if ('$value' in o) out.value = decodeValue(o.$value)
      return out
    }
    const rec: Record<string, unknown> = {}
    for (const k of Object.keys(o)) rec[k] = decodeValue(o[k])
    return rec
  }
  return v
}

/** Environment for the one-shot client, per seam target (UDS or TCP+mTLS). */
function clientEnv(target: SeamTarget, request: string, deadlineMs: number | null): NodeJS.ProcessEnv {
  const base: NodeJS.ProcessEnv = { ...process.env, BRIDGE_REQ: request }
  if (deadlineMs != null) base.BRIDGE_DEADLINE_MS = String(Math.max(1, Math.round(deadlineMs)))
  if (isNetwork(target)) {
    return {
      ...base,
      BRIDGE_MODE: 'tcp',
      BRIDGE_HOST: target.host,
      BRIDGE_PORT: String(target.port),
      BRIDGE_CERT: target.tls.cert,
      BRIDGE_KEY: target.tls.key,
      BRIDGE_CA: target.tls.ca,
      BRIDGE_SERVERNAME: target.tls.server_hostname ?? target.host,
    }
  }
  return { ...base, BRIDGE_MODE: 'uds', BRIDGE_SOCK: target }
}

/** One blocking seam round-trip to `target`. `deadlineMs` bounds the reply: a
 *  wedged provider surfaces as a `SeamDeadlineError` (a hang), a dropped
 *  connection as a plain `Error` (a death), a provider-side failure as the
 *  marshalled error — the same three disjoint fault kinds as the py client. */
function seamCall(
  target: SeamTarget,
  key: string,
  method: string,
  args: unknown[],
  deadlineMs: number | null,
): unknown {
  const request = JSON.stringify({ key, method, args: args.map(encodeBigInts) })
  let out: string
  try {
    out = execFileSync(process.execPath, ['-e', CLIENT_SRC], {
      env: clientEnv(target, request, deadlineMs),
      encoding: 'utf8',
      // Backstop the in-child deadline timer: if the child itself wedges, kill
      // it a little past the seam deadline and treat that as the same breach.
      timeout: deadlineMs != null ? Math.round(deadlineMs) + 5000 : undefined,
      maxBuffer: 64 * 1024 * 1024,
    })
  } catch (error) {
    if ((error as { code?: string }).code === 'ETIMEDOUT') {
      throw new SeamDeadlineError(key, method, deadlineMs ?? 0)
    }
    throw error
  }
  const reply = JSON.parse(out)
  if (reply.seamDeadline) throw new SeamDeadlineError(key, method, deadlineMs ?? 0)
  if (!reply.ok) throw new Error(reply.error)
  return decodeValue(reply.value)
}

/** Watch a provider for death: connect an idle monitor to `target` (UDS or
 *  TCP+mTLS) and call `onLost` once when the connection drops after having been
 *  established (peer death — the mTLS monitor EOFs exactly as the UDS one does).
 *  Returns a handle whose `close()` tears the monitor down without firing. */
export function watchPeer(
  target: SeamTarget,
  onLost: () => void,
): { close: () => void; ready: Promise<void> } {
  let fired = false
  let closed = false
  let everConnected = false
  let monitor: net.Socket | tls.TLSSocket
  let markReady: () => void
  const ready = new Promise<void>((resolve) => { markReady = resolve })
  const established = () => {
    everConnected = true
    markReady()
  }
  const connect = () => {
    if (closed) return
    if (isNetwork(target)) {
      monitor = tls.connect({
        host: target.host,
        port: target.port,
        ca: fs.readFileSync(target.tls.ca),
        cert: fs.readFileSync(target.tls.cert),
        key: fs.readFileSync(target.tls.key),
        servername: target.tls.server_hostname ?? target.host,
      })
      monitor.on('secureConnect', established)
    } else {
      monitor = net.connect(target)
      monitor.on('connect', established)
    }
    monitor.on('error', () => {}) // a 'close' event always follows
    monitor.on('close', () => {
      if (closed) return
      if (!everConnected) {
        setTimeout(connect, 50) // provider not up yet; keep trying
        return
      }
      if (fired) return
      fired = true
      onLost()
    })
  }
  connect()
  return {
    ready,
    close() {
      closed = true
      try {
        monitor?.destroy()
      } catch {
        /* best-effort */
      }
    },
  }
}

/** A cordis component that provides `key` via a proxy forwarding to `target`
 *  (a UDS path, or a `{ host, port, tls }` network TCP+mTLS endpoint), plus
 *  `onPeerLost` to observe the provider's death. `deadlineMs` bounds each seam
 *  call; on a **network** seam a breached deadline is treated as a lost peer and
 *  triggers reactive withdrawal (a wedged remote provider is, to a consumer,
 *  indistinguishable from a dead one — the seam is unusable either way), while a
 *  local UDS seam keeps the death-only withdrawal it always had. */
export function makeProxy(
  key: string,
  methods: string[],
  target: SeamTarget,
  deadlineMs: number | null = null,
) {
  const lostCallbacks: Array<() => void> = []
  let fired = false
  const fireLost = () => {
    if (fired) return
    fired = true
    for (const cb of lostCallbacks) cb()
  }
  const network = isNetwork(target)
  const monitor = watchPeer(target, fireLost)

  const proxy: Record<string, (...args: unknown[]) => unknown> = {}
  for (const method of methods) {
    proxy[method] = (...args: unknown[]) => {
      try {
        return seamCall(target, key, method, args, deadlineMs)
      } catch (error) {
        // A network seam that breaches its deadline withdraws: the remote
        // provider is unreachable-in-time, so the consumer stops depending on
        // it rather than re-attempting against a wedged machine.
        if (network && error instanceof SeamDeadlineError) fireLost()
        throw error
      }
    }
  }

  const component = {
    name: `${key[0].toUpperCase()}${key.slice(1)}Proxy`,
    inject: [] as string[],
    provide: [key],
    apply(ctx: Context) {
      ctx.effect(function* () {
        yield () => monitor.close()
        yield (ctx as any).provide(key, proxy)
      }, `${key}-proxy/body`)
    },
  }

  return {
    component,
    proxy,
    onPeerLost: (cb: () => void) => lostCallbacks.push(cb),
    monitorReady: monitor.ready,
    close: () => monitor.close(),
  }
}

/** Normalize `serve`'s exports to key -> allowed method names.
 *
 *  Two accepted forms, mirroring backends/python/bridge.py:
 *  - `{ key: [method, ...] }` — the *declared* form: the service's own
 *    operation list, read off the IR (src/revl/placement.py ships it in the
 *    process spec), so the stub admits exactly what the `service` declaration
 *    admits.
 *  - `string[]` — the legacy key-only form (demo/bridge_ts_adt.mts). With no
 *    declared list the allowlist is derived at dispatch time from the provided
 *    object's own function-valued properties: weaker (it trusts the object,
 *    not the interface) but still an allowlist. */
function exportTable(
  exports: string[] | Record<string, string[]>,
): Map<string, Set<string> | null> {
  const table = new Map<string, Set<string> | null>()
  if (Array.isArray(exports)) {
    for (const key of exports) table.set(key, null)
  } else {
    for (const [key, methods] of Object.entries(exports)) table.set(key, new Set(methods ?? []))
  }
  return table
}

function publicMethods(service: unknown): Set<string> {
  const names = new Set<string>()
  for (let o = service as Record<string, unknown> | null; o && o !== Object.prototype; o = Object.getPrototypeOf(o)) {
    for (const name of Object.getOwnPropertyNames(o)) {
      if (name === 'constructor' || name.startsWith('_')) continue
      try {
        if (typeof (service as Record<string, unknown>)[name] === 'function') names.add(name)
      } catch {
        /* a getter that throws is not a method */
      }
    }
  }
  return names
}

// --- what a failure may carry BACK across the seam (item 421 F5) -------------
//
// Mirror of `backends/python/bridge.py`'s `seam_failure` / `confidential.py`
// `redact_call_text`, spelled locally because this file is loaded standalone by
// the node consumer and imports nothing from the runtime.
//
// The consumer is on the other side of a trust boundary. A forward crossing into
// a declared `Secret[T]` receiver authorises disclosure TO THE RECEIVER; it does
// not authorise the error channel to perform the reverse crossing the checker
// refuses statically. And the trigger needs no author interpolation: a plain
// `store.get(token)` miss throws a message quoting the token. So every argument
// value this call was made with is scrubbed out of the host error text, while the
// error's TYPE and the sentence around it survive, so the reply is still worth
// reading, just without the caller's bytes in it.

/** The placeholder a caller's own argument is rendered as inside seam error
 *  text. Must equal `confidential.REDACTED_ARG` on the python tier: a polyglot
 *  seam should produce the SAME marker whichever tier answered. */
export const REDACTED_ARG = '<redacted:arg>'

// An argument shorter than this is left alone: below it a substring match is a
// coin flip against ordinary English and replacing it would shred the
// diagnostic for no confidentiality gain. Same bound as the python tier.
const MIN_MATCHABLE_ARG = 3

function argNeedles(value: unknown, into: Set<string>): void {
  if (value === null || value === undefined || typeof value === 'boolean') return
  if (typeof value === 'string') {
    if (value.length >= MIN_MATCHABLE_ARG) into.add(value)
    return
  }
  if (typeof value === 'number' || typeof value === 'bigint') {
    const form = String(value)
    if (form.length >= MIN_MATCHABLE_ARG) into.add(form)
    return
  }
  if (Array.isArray(value)) {
    for (const item of value) argNeedles(item, into)
    return
  }
  if (typeof value === 'object') {
    // Values only: a record's KEYS are field names the author wrote, not the
    // caller's data, and erasing them would destroy the diagnostic's shape.
    for (const item of Object.values(value as Record<string, unknown>)) argNeedles(item, into)
  }
}

/** The error text a provider-side failure is allowed to send back to the
 *  consumer, with this call's own argument values replaced by `REDACTED_ARG`.
 *  Longest needle first, so one that contains another leaves no tail behind. */
export function seamFailure(error: unknown, args: unknown[]): string {
  let text = String(error)
  const needles = new Set<string>()
  argNeedles(args ?? [], needles)
  for (const needle of [...needles].sort((a, b) => b.length - a.length)) {
    if (needle && text.includes(needle)) text = text.split(needle).join(REDACTED_ARG)
  }
  return text
}

/** Provider side: export a declared surface from `ctx` over a Unix socket, so
 *  another process can proxy it. Dispatches each JSON call to the committed
 *  view `ctx[key][method](...args)` — but only after checking BOTH halves of
 *  the request against the declaration: an unknown key and an unknown method
 *  are refused identically, so the seam is exactly the enumerable surface the
 *  service declares (G8). The reply carries the value (or the error). */
export async function serve(
  ctx: Context,
  exports: string[] | Record<string, string[]>,
  socketPath: string,
): Promise<net.Server> {
  const table = exportTable(exports)

  const server = net.createServer((sock) => {
    let buf = ''
    sock.on('data', async (chunk) => {
      buf += chunk
      let nl: number
      while ((nl = buf.indexOf('\n')) >= 0) {
        const line = buf.slice(0, nl)
        buf = buf.slice(nl + 1)
        if (!line) continue
        let reply: unknown
        // Hoisted so the catch can scrub the failing call's own arguments out of
        // the host error text (item 421 F5). A line that fails to parse never
        // assigns it, and an empty needle set leaves the text alone.
        let callArgs: unknown[] = []
        try {
          const req = JSON.parse(line)
          callArgs = req.args ?? []
          if (!table.has(req.key)) {
            reply = { ok: false, error: `key ${req.key} is not exported by this process` }
          } else {
            const service = (ctx as any)[req.key]
            const allowed = table.get(req.key) ?? publicMethods(service)
            if (!allowed.has(req.method)) {
              const listed = [...allowed].sort().join(', ') || '(none)'
              reply = {
                ok: false,
                error: `method ${req.method} is not exported for key ${req.key} (exported: ${listed})`,
              }
            } else {
              let result = service[req.method](...(req.args ?? []))
              if (result && typeof result.then === 'function') result = await result
              reply = { ok: true, value: encodeValue(result ?? null) }
            }
          }
        } catch (error) {
          reply = { ok: false, error: seamFailure(error, callArgs) }
        }
        sock.write(JSON.stringify(reply) + '\n')
      }
    })
    sock.on('error', () => {})
  })

  await new Promise<void>((resolve) => server.listen(socketPath, () => resolve()))
  return server
}

File: backends/go/placement_runner/bridge/bridge.go
// Package bridge is the cordis-go side of revl's cross-process interop bridge
// (docs/interop-bridge.md §3): a service provided in one OS process, consumed
// in another, over a Unix-domain socket.
//
// It is the transport half only — connection, newline-delimited JSON framing,
// the request/reply envelope, and peer-death monitoring. The value marshalling
// (records, ADTs, Result, Opt) is done by per-service code the emitter
// generates (backends/go/emit.py), because cordis-go services are static Go
// interfaces (like cordis-rs traits): a runtime-generic proxy is impossible, so
// generality is codegen, not reflection.
//
// The wire protocol is byte-compatible with backends/python/bridge.py: a
// request is {"key","method","args"} and a reply {"ok":true,"value":…} or
// {"ok":false,"error":…}, one JSON object per line. A Go proxy calling a Python
// stub (and vice-versa) interoperate because both speak exactly this envelope.
//
// Peer death is withdrawal (R2/R3): Monitor holds an idle second connection and
// fires its callback on EOF when the provider process goes away, so the runner
// can dispose the proxy fiber and let every dependent deactivate reactively —
// the same reactive path a local provider swap takes, now spanning processes.
package bridge

import (
	"bufio"
	"bytes"
	"encoding/json"
	"errors"
	"net"
	"sort"
	"strings"
	"sync"
	"time"

	"revl.goplacement/estop"
)

// dialAttempts / dialDelay mirror backends/python/bridge.py::_connect: under
// placement the provider and consumer start concurrently, so the socket may
// not exist yet; retrying makes start order irrelevant.
//
// Only while the PROCESS graph is acyclic. Two processes that each require a
// key the other provides both block in the dial and neither reaches its
// listen, so both die once the attempts run out. Nothing rejects that
// composition today; see docs/design/438-petri-reachability.md §5.2.
const (
	dialAttempts = 200
	dialDelay    = 50 * time.Millisecond
)

// Dial connects to a Unix socket, retrying while the provider comes up.
func Dial(path string) (net.Conn, error) {
	var last error
	for i := 0; i < dialAttempts; i++ {
		conn, err := net.Dial("unix", path)
		if err == nil {
			return conn, nil
		}
		last = err
		time.Sleep(dialDelay)
	}
	if last == nil {
		last = errors.New("bridge: could not connect to " + path)
	}
	return nil, last
}

// request is the consumer -> provider envelope. Field order (key, method, args)
// and JSON names match backends/python/bridge.py exactly.
type request struct {
	Key    string `json:"key"`
	Method string `json:"method"`
	Args   []any  `json:"args"`
}

// reply is the provider -> consumer envelope. `value` is decoded lazily by the
// generated proxy at the method's static return type; `error` carries a failure
// marshalled across the seam.
type reply struct {
	Ok    bool            `json:"ok"`
	Value json.RawMessage `json:"value"`
	Error string          `json:"error"`
}

// Client is one synchronous RPC connection to a provider stub. cordis-go
// provided methods are called synchronously, so a call is a blocking
// round-trip; peer death is observed on a separate Monitor connection.
type Client struct {
	mu   sync.Mutex
	conn net.Conn
	r    *bufio.Reader
}

// NewClient dials the stub at path (retrying) and returns a ready client.
func NewClient(path string) (*Client, error) {
	conn, err := Dial(path)
	if err != nil {
		return nil, err
	}
	return &Client{conn: conn, r: bufio.NewReader(conn)}, nil
}

// Call marshals one request, writes it, and returns the reply's raw `value`
// (or the remote error). `args` are already canonical-encoded values.
func (c *Client) Call(key, method string, args []any) (json.RawMessage, error) {
	// item 443 / issue #122 — the DISPATCH side of the E-Stop seam. Once an
	// operator arms the latch, this process stops DISPATCHING new crossings: the
	// outgoing call is REFUSED before it leaves the process, so nothing new
	// crosses the boundary. It returns an error rather than withdrawing the
	// proxy, because a halt is not a peer death: reactive withdrawal would
	// propagate a cooperative teardown to this proxy's dependents, which is
	// exactly the graceful unwind the E-Stop exists to avoid. The refused
	// caller's attempt lands in item 440's ambiguous tier, the designed outcome
	// of a halt (docs/design/443-estop.md). A go process is a consumer as well
	// as a provider, and the py design puts the check at every point that
	// dispatches OR accepts a crossing.
	if estop.EstopEngaged() {
		return nil, errors.New("revl E-Stop engaged: this process is HALTED and " +
			"refuses to dispatch new crossings (key " + key + ", method " + method +
			") — docs/design/443-estop.md")
	}
	if args == nil {
		args = []any{}
	}
	line, err := json.Marshal(request{Key: key, Method: method, Args: args})
	if err != nil {
		return nil, err
	}
	line = append(line, '\n')

	// Record the crossing as in flight for its round-trip: a crossing still out
	// when the latch trips is the AMBIGUOUS one the halt inventory names.
	seq := estop.BeginCrossing(key, method, "dispatch")
	defer estop.EndCrossing(seq)

	c.mu.Lock()
	defer c.mu.Unlock()
	if _, err := c.conn.Write(line); err != nil {
		return nil, err
	}
	respLine, err := c.r.ReadBytes('\n')
	if err != nil {
		return nil, errors.New("bridge peer closed the connection")
	}
	var rep reply
	if err := json.Unmarshal(respLine, &rep); err != nil {
		return nil, err
	}
	if !rep.Ok {
		if rep.Error == "" {
			rep.Error = "remote error"
		}
		return nil, errors.New(rep.Error)
	}
	return rep.Value, nil
}

// Close tears the RPC transport down.
func (c *Client) Close() error {
	c.mu.Lock()
	defer c.mu.Unlock()
	return c.conn.Close()
}

// Monitor dials a second, idle connection to path and calls onLost exactly once
// when the provider dies (the connection hits EOF). The monitor connection
// never sends: the stub's read loop simply sees EOF when the provider process
// (and its listener) go away.
func Monitor(path string, onLost func()) {
	conn, err := Dial(path)
	if err != nil {
		onLost()
		return
	}
	go func() {
		defer conn.Close()
		buf := make([]byte, 64)
		for {
			if _, err := conn.Read(buf); err != nil {
				break // EOF or error: the provider is gone
			}
		}
		onLost()
	}()
}

// Invoke dispatches one incoming call to a local service and returns the value
// to marshal back (or an error to report across the seam). The generated
// RevlInvoke has this shape.
type Invoke func(key, method string, args []json.RawMessage) (any, error)

// okReply / errReply are the provider's two reply shapes. Kept as explicit
// structs (not maps) so the JSON is deterministic and matches the reference.
type okReply struct {
	Ok    bool `json:"ok"`
	Value any  `json:"value"`
}

type errReply struct {
	Ok    bool   `json:"ok"`
	Error string `json:"error"`
}

// --- what a failure may carry BACK across the seam (roadmap item 421 F5) -----
//
// Mirror of backends/python/bridge.py's seam_failure and the same helper in
// backends/typescript/bridge.ts. The consumer is on the other side of a trust
// boundary: a forward crossing into a declared Secret[T] receiver authorises
// disclosure TO THE RECEIVER, and does not authorise the error channel to
// perform the reverse crossing the checker refuses statically. The trigger needs
// no author interpolation - a plain map miss produces a message quoting the key.
//
// So every argument value the call was made with is scrubbed out of the host
// error text, while the sentence around it survives: the reply still says what
// went wrong, just without the caller's bytes in it.

// RedactedArg is the placeholder a caller's own argument becomes inside seam
// error text. Must equal confidential.REDACTED_ARG on the python tier, so a
// polyglot seam produces the SAME marker whichever tier answered.
const RedactedArg = "<redacted:arg>"

// minMatchableArg is the length below which an argument is left alone: a shorter
// substring match is a coin flip against ordinary English and replacing it would
// shred the diagnostic for no confidentiality gain. Same bound as python and ts.
const minMatchableArg = 3

// argNeedles collects the string forms a decoded argument can take inside host
// error text. Booleans and null are skipped (their renderings are ordinary
// words); a record's KEYS are skipped too, because they are field names the
// author wrote rather than the caller's data.
func argNeedles(value any, into map[string]struct{}) {
	switch v := value.(type) {
	case string:
		if len(v) >= minMatchableArg {
			into[v] = struct{}{}
		}
	case json.Number:
		if len(v.String()) >= minMatchableArg {
			into[v.String()] = struct{}{}
		}
	case []any:
		for _, item := range v {
			argNeedles(item, into)
		}
	case map[string]any:
		for _, item := range v {
			argNeedles(item, into)
		}
	}
}

// SeamFailure is the error text a provider-side failure is allowed to send back
// to the consumer, with this call's own argument values replaced by RedactedArg.
// Longest needle first, so one that contains another leaves no tail behind.
func SeamFailure(err error, args []json.RawMessage) string {
	text := err.Error()
	needles := map[string]struct{}{}
	for _, raw := range args {
		decoder := json.NewDecoder(bytes.NewReader(raw))
		// UseNumber keeps an integer argument spelled the way the wire spelled
		// it, so 8675309 is matched as "8675309" and not as "8.675309e+06".
		decoder.UseNumber()
		var value any
		if decoder.Decode(&value) == nil {
			argNeedles(value, needles)
		}
	}
	ordered := make([]string, 0, len(needles))
	for needle := range needles {
		ordered = append(ordered, needle)
	}
	sort.Slice(ordered, func(i, j int) bool { return len(ordered[i]) > len(ordered[j]) })
	for _, needle := range ordered {
		text = strings.ReplaceAll(text, needle, RedactedArg)
	}
	return text
}

// Serve accepts connections on ln and answers each request via invoke until ln
// is closed. Each connection is handled on its own goroutine.
func Serve(ln net.Listener, invoke Invoke) {
	for {
		conn, err := ln.Accept()
		if err != nil {
			return // listener closed: process is shutting down
		}
		go serveConn(conn, invoke)
	}
}

func serveConn(conn net.Conn, invoke Invoke) {
	defer conn.Close()
	r := bufio.NewReader(conn)
	for {
		line, err := r.ReadBytes('\n')
		if len(line) > 0 {
			var req struct {
				Key    string            `json:"key"`
				Method string            `json:"method"`
				Args   []json.RawMessage `json:"args"`
			}
			if json.Unmarshal(line, &req) == nil {
				var out []byte
				if estop.EstopEngaged() {
					// item 443 / issue #122 — the ACCEPT side of the E-Stop seam.
					// An armed latch means an operator hit the button, so this
					// crossing is REFUSED before the service method runs: nothing
					// new crosses the boundary. The reply is an error, not a
					// value, and (unlike a cooperative teardown) no inverse is
					// replayed and nothing is discharged — the caller's attempt
					// lands in item 440's ambiguous tier, the designed outcome of
					// a halt (docs/design/443-estop.md).
					out, _ = json.Marshal(errReply{Ok: false, Error: "revl E-Stop engaged: " +
						"this process is HALTED and refuses new crossings (key " + req.Key +
						", method " + req.Method + ") — docs/design/443-estop.md"})
					out = append(out, '\n')
					if _, werr := conn.Write(out); werr != nil {
						return
					}
					if err != nil {
						return
					}
					continue
				}
				// Record the crossing as in flight WHILE its handler runs: a
				// crossing still executing when the latch trips is the AMBIGUOUS
				// one the halt inventory names (item 440). Cleared in a deferred
				// call so a panicking handler still leaves the registry clean.
				value, ierr := func() (any, error) {
					seq := estop.BeginCrossing(req.Key, req.Method, "accept")
					defer estop.EndCrossing(seq)
					return invoke(req.Key, req.Method, req.Args)
				}()
				if ierr != nil {
					// item 421 F5: never hand the consumer back the values it
					// called with.
					out, _ = json.Marshal(errReply{Ok: false, Error: SeamFailure(ierr, req.Args)})
				} else {
					out, _ = json.Marshal(okReply{Ok: true, Value: value})
				}
				out = append(out, '\n')
				if _, werr := conn.Write(out); werr != nil {
					return
				}
			}
		}
		if err != nil {
			return // EOF (monitor connection) or a read error
		}
	}
}
